[SR-12605] #sourceLocation doesn't allow underscores in line numbers

|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-12605      |
   |Radar             | rdar://problem/62202395         |
   |Original Reporter | @beccadax      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

Near-master, 15 April 2019


</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Bug, Parser, StarterBug        |
   |Assignee          | matthewspear (JIRA)      |
   |Priority          | Medium      |

   

   md5: 41076a6957823a7ff8a7fce6be55225d

  </details>





**Issue Description:**


``` none
  1> #sourceLocation(file: "a.swift", line: 1_000)
error: /var/folders/mw/1rhv6fn96gb8swwkt8zycwph0000gn/T/repl1-101710..swift:2:40: error: expected starting line number for #sourceLocation directive
#sourceLocation(file: "a.swift", line: 1_000)
                                       ^
  1> #sourceLocation(file: "a.swift", line: 1000)
  2> 1_000
$R0: Int = 1000
  3>  
```

Underscore separators in `#sourceLocation` directives probably ought to be supported.

--- a/include/swift/Parse/Parser.h
+++ b/include/swift/Parse/Parser.h
@@ -1653,6 +1653,8 @@ public:
   ParserResult<Expr> parseExprRegexLiteral();
 
   StringRef copyAndStripUnderscores(StringRef text);
+  StringRef stripUnderscoresIfNeeded(StringRef text,
+                                     SmallVectorImpl<char> &buffer);
 
   ParserStatus parseStringSegments(SmallVectorImpl<Lexer::StringSegment> &Segments,
                                    Token EntireTok,

--- a/lib/Parse/ParseDecl.cpp
+++ b/lib/Parse/ParseDecl.cpp
@@ -5809,7 +5809,9 @@ ParserStatus Parser::parseLineDirective(bool isLine) {
         diagnose(Tok, diag::expected_line_directive_number);
         return makeParserError();
       }
-      if (Tok.getText().getAsInteger(0, StartLine)) {
+      SmallString<16> buffer;
+      auto text = stripUnderscoresIfNeeded(Tok.getText(), buffer);
+      if (text.getAsInteger(0, StartLine)) {
         diagnose(Tok, diag::expected_line_directive_number);
         return makeParserError();
       }
@@ -5840,7 +5842,9 @@ ParserStatus Parser::parseLineDirective(bool isLine) {
       diagnose(Tok, diag::expected_line_directive_number);
       return makeParserError();
     }
-    if (Tok.getText().getAsInteger(0, StartLine)) {
+    SmallString<16> buffer;
+    auto text = stripUnderscoresIfNeeded(Tok.getText(), buffer);
+    if (text.getAsInteger(0, StartLine)) {
       diagnose(Tok, diag::expected_line_directive_number);
       return makeParserError();
     }

--- a/lib/Parse/ParseExpr.cpp
+++ b/lib/Parse/ParseExpr.cpp
@@ -1009,6 +1009,17 @@ StringRef Parser::copyAndStripUnderscores(StringRef orig) {
   return StringRef(start, p - start);
 }
 
+StringRef Parser::stripUnderscoresIfNeeded(StringRef text,
+                                           SmallVectorImpl<char> &buffer) {
+  if (text.contains('_')) {
+    buffer.clear();
+    llvm::copy_if(text, std::back_inserter(buffer),
+                  [](char ch) { return ch != '_'; });
+    return StringRef(buffer.data(), buffer.size());
+  }
+  return text;
+}
+
 /// Disambiguate the parse after '{' token that is in a place that might be
 /// the start of a trailing closure, or start the variable accessor block.
 ///

--- a/test/Parse/line-directive.swift
+++ b/test/Parse/line-directive.swift
@@ -67,3 +67,13 @@ enum E {
 // CHECK: sr8772.swift:400:2: error: expected member name following '.'
 // CHECK: sr8772.swift:400:3: error: consecutive statements on a line must be separated by ';'
 // CHECK: sr8772.swift:400:3: error: expected expression
+
+// https://github.com/apple/swift/issues/55049
+class I55049 {
+#sourceLocation(file: "issue-55049.swift", line: 1_000)
+    let bar = 12
+#sourceLocation(file: "issue-55049.swift", line: 2_000)
+}
+
+#line 1_000 "issue-55049.swift"
+class I55049_1 {}

